atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java

```diff
@@ -1666,48 +1676,59 @@ public void putMetadataForTables(final Map<TableReference, byte[]> tableRefToMet
+                tableRefToMetadata, (tableRef, metadata) -> CassandraKeyValueServices.getOldMetadataCell(tableRef));
 
-        // technically we're racing other services from here on, during an update period,
+        // technically we're racing other nodes from here on, during an update period,
         // but the penalty for not caring is just some superfluous schema mutations and a
         // few dead rows in the metadata table.
-        Map<Cell, Value> persistedMetadata = get(
-                AtlasDbConstants.DEFAULT_METADATA_TABLE,
-                requestForLatestDbSideMetadata);
+        Map<Cell, Value> existingMetadataAtNewName = get(AtlasDbConstants.DEFAULT_METADATA_TABLE,
+                tableRefToNewCell.values().stream()
+                        .collect(Collectors.toMap(Functions.identity(), Functions.constant(Long.MAX_VALUE))));
+
+        Map<Cell, Value> existingMetadataAtOldName = get(AtlasDbConstants.DEFAULT_METADATA_TABLE,
+                tableRefToOldCell.values().stream()
+                        .collect(Collectors.toMap(Functions.identity(), Functions.constant(Long.MAX_VALUE))));
+
         final Map<Cell, byte[]> updatedMetadata = Maps.newHashMap();
-        final Collection<CfDef> updatedCfs = Lists.newArrayList();
-        for (Entry<Cell, byte[]> entry : metadataRequestedForUpdate.entrySet()) {
-            if (updatedMetadataFound(persistedMetadata.get(entry.getKey()), entry.getValue())) {
-                updatedMetadata.put(entry.getKey(), entry.getValue());
-                updatedCfs.add(getCfForTable(
-                        CassandraKeyValueServices.tableReferenceFromBytes(entry.getKey().getRowName()),
-                        entry.getValue(),
-                        config.gcGraceSeconds()));
+        final Set<CfDef> updatedCfs = Sets.newHashSet();
+
+        tableRefToNewCell.forEach((tableRef, newCell) -> {
+            if (existingMetadataAtNewName.containsKey(newCell)) {
+                if (metadataIsDifferent(
+                        existingMetadataAtNewName.get(newCell).getContents(), tableRefToMetadata.get(tableRef))) {
+                    // found existing metadata at new name, but we're performing an update
+                    updatedMetadata.put(newCell, tableRefToMetadata.get(tableRef));
+                    updatedCfs.add(getCfForTable(tableRef, tableRefToMetadata.get(tableRef), config.gcGraceSeconds()));
+                }
+            } else if (existingMetadataAtOldName.containsKey(tableRefToOldCell.get(tableRef))) {
+                if (metadataIsDifferent(
+                        existingMetadataAtOldName.get(tableRefToOldCell.get(tableRef)).getContents(),
+                        tableRefToMetadata.get(tableRef))) {
+                    // found existing metadata at old name, but we're performing an update
+                    updatedMetadata.put(tableRefToOldCell.get(tableRef), tableRefToMetadata.get(tableRef));
```

Comment:
So we keep the old name in cases where it's used?

Reply:
Yep, we continue to use the old name, in read, update, and delete cases.
The main improvements are thus for {existing deployments with old tables that were always lowercased} and {old and new deployments who will just write out all future tables more nicely}.

The negative is a light amount of legacy tech debt and bonus complexity baked into CKVS.

Reply:
I will always 👍 bonus complexity